### PR TITLE
Serialize/Deserialize source locations for instructions

### PR DESCRIPTION
**Pull request by usama54321**
This PR serializes/deserializes source locations for instructions.

### CODE REVIEW

**Comment by usama54321** [comment]
@swift-ci test

**Comment by usama54321** [comment]
@swift-ci test

**Comment by usama54321** [comment]
@swift-ci test

**Comment by usama54321** [comment]
@swift-ci test

**Comment by usama54321** [comment]
@swift-ci test

File: lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp [hunk 1]
```diff
@@ -994,6 +994,9 @@ static bool diagnoseUnreachableBlock(
     if (Loc.is<ImplicitReturnLocation>() || Loc.isAutoGenerated())
       return false;
 
+    if (Loc.isFilenameAndLocation())
```
Comment (eeckstein) [inline lib/SILOptimizer/Mandatory/DiagnoseUnreachable.cpp:997]: Can you please add a comment why this is done?

**Comment by usama54321** [comment]
@swift-ci test

**Comment by usama54321** [comment]
@swift-ci test

File: include/swift/SIL/SILLocation.h [hunk 2]
```diff
@@ -552,6 +558,8 @@ class RegularLocation : public SILLocation {
       : SILLocation(L, RegularKind, Implicit) {}
   RegularLocation(FilenameAndLocation *filePos)
     : SILLocation(filePos, RegularKind) {}
+  RegularLocation(FilenameAndLocation *filePos, bool Implicit)
```
Comment (adrian-prantl) [inline include/swift/SIL/SILLocation.h:561]: We should consider merging this with the function above and either update the call sites to explicitly state "explicit" or have a default argument.

File: include/swift/SIL/SILLocation.h [hunk 3]
```diff
@@ -622,6 +630,9 @@ class ReturnLocation : public SILLocation {
   ReturnLocation(FilenameAndLocation *filePos)
     : SILLocation(filePos, ReturnKind) {}
 
+  ReturnLocation(FilenameAndLocation *filePos, bool Implicit)
```
Comment (adrian-prantl) [inline include/swift/SIL/SILLocation.h:633]: same here

File: lib/Serialization/DeserializeSIL.cpp [hunk 4]
```diff
@@ -394,6 +394,41 @@ SILDeserializer::readNextRecord(SmallVectorImpl<uint64_t> &scratch) {
...
+    FNameLoc = ParsedLocs[LocID];
+  } else {
+    ValueID Row = 0, Col = 0, FNameID = 0;
+    SourceLocLayout::readRecord(scratch, Row, Col, FNameID, LocationKind);
+
+
+    FNameLoc = SILLocation::FilenameAndLocation::alloc(
+        Row, Col, MF->getIdentifierText(FNameID), SILMod);
+
+    ParsedLocs.insert({ParsedLocs.size() + 1, FNameLoc});
```
Comment (adrian-prantl) [inline lib/Serialization/DeserializeSIL.cpp:415]: Is this a DenseMap of consecutive Int keys? If yes, why can't this a vector?

File: lib/Serialization/SerializeSIL.cpp [hunk 5]
```diff
@@ -3074,6 +3083,47 @@ void SILSerializer::writeSILProperty(const SILProperty &prop) {
...
+  auto OpaquePtr = SLoc.getOpaquePointerValue();
+  uint8_t LocationKind;
+  switch(Loc.getKind()) {
+    case SILLocation::ReturnKind:
+      LocationKind = SILLocation::ReturnKind;
+      break;
+    case SILLocation::ImplicitReturnKind:
+      LocationKind = SILLocation::ImplicitReturnKind;
+      break;
+    //TODO: handle other cases correctly
```
Comment (adrian-prantl) [inline lib/Serialization/SerializeSIL.cpp:3097]: maybe explicitly list all the cases so we get a warning when a new one is added.

**Review by adrian-prantl — COMMENTED** [review]
I only found a few minor comments.

File: lib/Serialization/DeserializeSIL.cpp [hunk 6]
```diff
@@ -394,6 +394,41 @@ SILDeserializer::readNextRecord(SmallVectorImpl<uint64_t> &scratch) {
   return maybeKind;
 }
 
+std::optional<SILLocation> SILDeserializer::readLoc(unsigned kind,
+                                     SmallVectorImpl<uint64_t> &scratch) {
+  unsigned LocationKind = 0;
+  SILLocation::FilenameAndLocation *FNameLoc = nullptr;
+  if (kind == SIL_SOURCE_LOC_REF) {
```
Comment (eeckstein) [inline lib/Serialization/DeserializeSIL.cpp:401]: Can you add a short comment explaining the strategy of serializing a sourceloc only the first time it appears and referencing this first sourceloc for successive appearances.

**Review by eeckstein — COMMENTED** [review]
Can you please update the comment for `struct FilenameAndLocation`:
```
  /// In future we might also use it for de-serialized locations from a
  /// swiftmodule file.
```
The future is now 🙂

**Comment by usama54321** [comment]
@swift-ci test

File: lib/SIL/IR/SILLocation.cpp [hunk 7]
```diff
@@ -137,6 +137,8 @@ SourceLoc SILLocation::getStartSourceLoc() const {
     return SourceLoc();
   if (isSILFile())
     return storage.sourceLoc;
+  if (getStorageKind() == FilenameAndLocationKind)
+    return SourceLoc();
```
Comment (adrian-prantl) [inline lib/SIL/IR/SILLocation.cpp:141]: This looks like it's working around incomplete serialization?

**Review by adrian-prantl — APPROVED** [review]

**Comment by usama54321** [comment]
@swift-ci test

**Pull request merged**